generator/src/lib.rs

```diff
@@ -63,83 +69,89 @@ impl quote::ToTokens for CType {
-            alt!(tag!("0x") | tag!("0X")),
-            map!(
-                pair!(hex_digit1, ctype),
-                |(num, typ)| (typ, format!("0x{}{}", num.to_ascii_lowercase(), typ.to_string())
-            )
-        )
-    )
-);
-
-named!(inverse_number<&str, (CType, String)>,
-    map!(
-        delimited!(
-            tag!("("),
-            pair!(
-                preceded!(tag!("~"), decimal_number),
-                opt!(preceded!(tag!("-"), digit1))
-            ),
-            tag!(")")
+fn parse_cexpr<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, (CType, String), E> {
+    (alt((
+        map(parse_cfloat, |f| (CType::Float, format!("{:.2}", f))),
+        parse_inverse_number,
+        parse_decimal_number,
+        parse_hexadecimal_number,
+    )))(i)
+}
+
+fn parse_cfloat<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, f32, E> {
+    (terminated(nom::number::complete::float, one_of("fF")))(i)
+}
+
+fn parse_inverse_number<'a, E: ParseError<&'a str>>(
+    i: &'a str,
+) -> IResult<&'a str, (CType, String), E> {
+    (map(
+        delimited(
+            tag("("),
+            tuple((
+                preceded(tag("~"), parse_decimal_number),
+                opt(preceded(tag("-"), digit1)),
+            )),
+            tag(")"),
         ),
         |((ctyp, num), minus_num)| {
             let expr = if let Some(minus) = minus_num {
                 format!("!{}-{}", num, minus)
-            }
-            else{
+            } else {
                 format!("!{}", num)
             };
             (ctyp, expr)
-        }
-    )
-);
-
-named!(cfloat<&str, f32>,
-    terminated!(nom::number::complete::float, one_of!("fF"))
-);
+        },
+    ))(i)
+}
 
 // Like a C string, but does not support quote escaping and expects at least one character.
 // If needed, use https://github.com/Geal/nom/blob/8e09f0c3029d32421b5b69fb798cef6855d0c8df/tests/json.rs#L61-L81
-named!(c_include_string<&str, String>,
-    delimited!(
-        char!('"'),
-        map!(
-            many1!(none_of!("\"")),
-            |chars| chars.iter().map(char::to_string).join("")
-        ),
-        char!('"')
-    )
-);
-
-named!(c_include<&str, String>,
-    preceded!(tag!("#include"), preceded!(multispace1, c_include_string))
-);
+fn parse_c_include_string<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, String, E> {
+    (delimited(
+        char('"'),
+        map(many1(none_of("\"")), |c| {
+            c.iter().map(char::to_string).join("")
+        }),
+        char('"'),
+    ))(i)
+}
+
+fn parse_c_include<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, String, E> {
+    (preceded(
+        tag("#include"),
+        preceded(multispace1, parse_c_include_string),
+    ))(i)
+}
+
+fn parse_decimal_number<'a, E: ParseError<&'a str>>(
+    i: &'a str,
+) -> IResult<&'a str, (CType, String), E> {
+    (map(tuple((digit1, parse_ctype)), |(dig, ctype)| {
+        (ctype, dig.to_string())
+    }))(i)
```

Comment:
Perhaps you could reduce the scope of this `map` to apply to `digit1` only? Something like `tuple(map(digit1, u32::to_string), parse_ctype)` or something akin to that?

Reply:
Turns out:

- `digit1` (obviously :facepalm:) parses to `str`, so that'd be `str::to_string`;
- This `map` is also used to invert the argument.

I think for the 2-tuple case we can use `nom`'s `pair()` parser, and I'd be totally happy to use `str::to_string` or `str::to_owned` here so that the source type is a bit more clear.